### DISCUSSION TITLE
[Feature] When no exif timestamp, select filename as timestamp source if corresponding to a pattern

### DISCUSSION THREAD

**Discussion by Chuckame** [immich-app/immich#6261] [Feature Request]
## Issue
I have a lot of photos coming from whatsapp, and they are not storing exif data.

Because of that miss, immich fallback on the file creation date.

## Feature request
It would be nice to fallback first on the file name if it respect a pattern (that is done by a lot of apps, like facebook, whatsapp, android camera, ...). Also, having multiple patterns would be awesome, even more if it's customizable (env vars at least).

Here is for whatsapp a regex as an example: `(VID|IMG)-(?<year>\d{4})(?<month>\d{2})(?<day>\d{2})-WA\d+`. That will match and retrieve the `year`, `month` and `day`. If not matching, then fallback as usual using the file creation date.

_I did not find an explicit feature request of this. Feel free to close/link to other discussions_

### Platform

- [X] Server
- [ ] Web
- [ ] Mobile

**Comment by Sums123** [comment]
Hi, not sure if you are already aware, but until Immich implements such a feature, exiftool has proven helpful to me
https://exiftool.org/faq.html

`exiftool "-alldates<filename" c:\images`

**Reply by Chuckame** [reply]
OH, really? I'll look at it! Thanks

**Comment by unhighghlow** [comment]
+1. Would be really nice to have

**Comment by unhighghlow** [comment]
It would also be really useful if you could manually specify the date-matching regex or multiple regexes, which the server could try sequentially for each asset missing a date. (Or manually triggered)

**Comment by pulpog** [comment]
There was a lot of discussion regarding this in now-closed duplicate (#9116).
This is actually a generic problem with any photos lacking exif, regardless of their source. iPhones further do not save WhatsApp images with a datestamp filename (instead with a UUID) so this would be a partial solution. At least with iPhone, the saved date is available to apps from the device timeline.

**Reply by Ferdi265** [reply]
I think recovering dates from device-specific sources (file modification time, iPhone/Android device media timeline APIs) and recovering dates from filenames are two different problems that should both be addressed.

This will always be a partial solution as in some cases there genuinely is no date information or the information is encoded in e.g. the filename in a nonstandard or undetectable way (unix timestamp, different epoch, american va european month/day day/month order). This does not mean it is not worth trying to improve the situation.

My opinions:
- The mobile app(s) should gain the ability to recover date information from the device media timeline and provide it to immich (e.g. via metadata provided via the upload API?)
- The server should gain the ability to recover date information from the filename using configurable matching patterns (and some common defaults like signal-YYYY-MM-DD-HH-MM-SS or IMG-YYYYMMDD-WA)
- There should be server or user configuration about which information to prefer in case of conflicts or missing information, so users or admins can set their preferred order
  - reasonable default: EXIF > filename pattern > device timeline
